hare/eligibility/oracle.go

```diff
@@ -229,14 +231,20 @@ func (o *Oracle) actives(layer types.LayerID) (map[string]struct{}, error) {
 		return nil, errGenesis
 	}
 
+	// lock until any return
+	// note: no need to lock per safeEp - we do not expect many concurrent requests per safeEp (max two)
+	o.lock.Lock()
+
 	// check cache
-	if val, exist := o.cache.Get(sl); exist {
+	if val, exist := o.cache.Get(safeEp); exist {
+		o.lock.Unlock()
 		return val.(map[string]struct{}), nil
 	}
 
 	// build a map of all blocks on the current layer
 	mp, err := o.blocksProvider.ContextuallyValidBlock(sl)
```

Comment:
not sure how hard is this and if it has to be protected

Reply:
not very hard its basically 2 database fetches for each call to ContextuallyValidBlock which is called per layer
no calculations

Reply:
Good, leaving it as is.